Fix various errors found during LTO build

Thaodan opened the pull request:
Also fix out of tree builds when trying to find `scratchbox2_version.h`

Contains #22.

Thaodan committed:
[sb2] Avoid errors in out of tree builds for files from $(OBJDIR)

Previously in some instance the assumption was $(SRCDIR) = $(OBJDIR).
If $(OBJDIR) isn't equal to $(SRCDIR) these files were not found anymore.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Avoid errors in out of tree builds for files from $(OBJDIR)

Previously in some instance the assumption was $(SRCDIR) = $(OBJDIR).
If $(OBJDIR) isn't equal to $(SRCDIR) these files were not found anymore.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[utils] Use va_args in open_nomap_nolog to match wrappers. JB#61866

Resolves error when using a compiler with stricter settings such as
when using LTO.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[utils] Use va_args in open_nomap_nolog to match wrappers. JB#61866

Resolves error when using a compiler with stricter settings such as
when using LTO.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[preload] Adapt glibc in open ignore 'mode' with O_TMPFILE in flags. JB#61866

Both open and openat load their last argument 'mode' lazily, using
va_arg() only if O_CREAT is found in oflag. This is wrong, mode is also
necessary if O_TMPFILE is in oflag.

Adapt __OPEN_NEEDS_MODE(oflag) present since glibc 2.22 to solve the
issue.

Read more down below:
https://sourceware.org/git/?p=glibc.git;a=commit;h=65f6f938cd562a614a68e15d0581a34b177ec29d

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[preload] Adapt glibc in open ignore 'mode' with O_TMPFILE in flags. JB#61866

Both open and openat load their last argument 'mode' lazily, using
va_arg() only if O_CREAT is found in oflag. This is wrong, mode is also
necessary if O_TMPFILE is in oflag.

Adapt __OPEN_NEEDS_MODE(oflag) present since glibc 2.22 to solve the
issue.

Read more down below:
https://sourceware.org/git/?p=glibc.git;a=commit;h=65f6f938cd562a614a68e15d0581a34b177ec29d

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Don't run make twice

Once in autogen.sh and once in %make

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Don't run make twice

Once in autogen.sh and once in %make

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[preload] Fix generated dependency order

Previously preload/exported.h depended on preload/wrappers.c even thou
both where generated from interface.master by gen-interface.pl.

Refactor so that all generated preloader files depend on the interface
and the gen-interface script.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[preload] Fix generated dependency order

Previously preload/exported.h depended on preload/wrappers.c even thou
both where generated from interface.master by gen-interface.pl.

Refactor so that all generated preloader files depend on the interface
and the gen-interface script.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[llbuild] Refactor import_dir to match other Makefiles

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[llbuild] Refactor import_dir to match other Makefiles

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

martyone commented on execs/exec_map_script_interp.c:
```diff
@@ -5,11 +5,13 @@
 
-#include "mapping.h"
-#include "sb2.h"
+#include <config.h>
+
+#include <mapping.h>
+#include <sb2.h>
+
```
Wouldn't it be more correct to have all files that are located in source/build tree referred to with double-quoted paths? Their use make it immediately clear that the files come from the source/build tree (no matter which subdir) and not from system paths. I see that the original sources are not 100% consistent in that, but double quotes are used way more often.

Thaodan replied:
The difference is between files which come from the include path and which come from the source/build directory.  These use the include-path -I, besides to clearly indicate which files come from the include-path and which not this also ensures that they won't be taken from other directories such as the directory relative to the source file.

Check below for more:
https://www.gnu.org/software/emacs/manual/html_node/autoconf/Configuration-Headers.html

martyone replied:
yeah it's just the lack of any clue on where the headers with possibly very generic name come from, that makes me think about these..

Thaodan replied:
Any ideas how to make this more clear? After reading the manuals it is quite clear but this might be a detail easy to oversee for those not in the know.

martyone commented on execs/exec_map_script_interp.c:
```diff
@@ -5,11 +5,13 @@
 #include "libsb2.h"
 #include "exported.h"
-
 #include "sb2_execs.h"
```
This separation of includes located in the same directory seems to be consistently followed in the original source, so I would rather keep it.

martyone commented on preload/Makefile:
```diff
@@ -44,6 +44,11 @@ targets := $(targets) $(D)/libsb2.$(SHLIBEXT)
+# without
+.NOTPARALLEL: $(OBJDIR)/preload/exported.h
+.NOTPARALLEL: preload/exported.h
 $(OBJDIR)/preload/wrappers.c: preload/interface.master preload/gen-interface.pl
```
This does not look like a right use of `.NOTPARALLEL` according to the docs.

I think the issue is caused by the way the dependencies are expressed. `exported.h` can hardly depend on `wrappers.c` when both are generated with `gen-interface.pl` based on `interface.master`. Depending on the order in which they are generated, `gen-interface.pl` may get invoked repeatedly and possibly in parallel. Maybe it would be enough to move the target `exported.h` from line 46 to line 52 (47 in the original file).

```suggestion
$(D)/ldexportlist: $(OBJDIR)/preload/wrappers.c
$(OBJDIR)/preload/exported.h $(OBJDIR)/preload/wrappers.c: preload/interface.master preload/gen-interface.pl
```

Thaodan replied:
Hm I have to look closer first. I will check and comeback.

Thaodan replied:
Good call. Should be better now.  While investigating also also found an earlier change related to this which wasn't fully done and is now fixed.

martyone commented on llbuild/Makefile.include:
```diff
@@ -41,15 +41,17 @@ endef
+	objs-y := $(join $(subst ./, $(D)/, $(dir $(objs))), $(notdir $(objs)))
+	objs-all := $(objs-all) \
+	 	$(patsubst %,$(OBJDIR)/%,$(objs-y))
+    deps-all := $(deps-all) $(patsubst %,$(OBJDIR)/%,$(deps))
+    include $(wildcard $(deps))
```
Wrong indent?

Thaodan committed:
preload: Remove mentions of ldexportlist or replace with export.map

ldexportlist was removed earlier for export.map. Replace with
export.map where not done already. Remove obsolete references such as
the target in preload/Makefile.

Fixes: 20180ac890cd54ccc1ee1b450a660681e87ce1bd
Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
preload: Remove mentions of ldexportlist or replace with export.map

ldexportlist was removed earlier for export.map. Replace with
export.map where not done already. Remove obsolete references such as
the target in preload/Makefile.

Fixes: 20180ac890cd54ccc1ee1b450a660681e87ce1bd
Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

mlehtima commented on preload/Makefile:
```diff
@@ -42,31 +43,42 @@ $(D)/libsb2.$(SHLIBEXT): LIBS := -ldl $(LIBSB2_LIBS)
+.PRECIOUS: $(prelader_generated_public) $(prelader_generated_private)
 
-CLEAN_FILES += $(generated)
+CLEAN_FILES += $(prelader_generated_public) $(prelader_generated_private)
```
Typos in this and above, prelader_generated_* should be preloader_generated_*

Thaodan replied:
oh fu thanks
